src/logics/syncswap/utils.ts

```diff
@@ -0,0 +1,688 @@
+        if (poolA && poolBase && poolB) {
+          paths.push(
+            createPath([
+              { routePool: poolA, tokenIn },
+              { routePool: poolBase, tokenIn: baseToken1 },
+              { routePool: poolB, tokenIn: baseToken2 },
+            ])
+          );
+        }
+      }
+    }
+  }
+
+  return paths;
+}
+
+export function findAllPossiblePaths(tokenIn: string, tokenOut: string, routePools: RoutePools, baseTokens: string[]) {
+  const { poolsDirect, poolsA, poolsB, poolsBase } = routePools;
+
+  const paths: Path[] = [];
+
+  // Direct pools
+  for (const routePool of poolsDirect) {
+    if (hasLiquidity(routePool, tokenIn)) {
+      const path = createPath([{ routePool, tokenIn }]);
+      paths.push(path);
+    }
+  }
+
+  const routePoolMap: Map<string, RoutePool> = new Map();
+  for (const routePool of [...poolsA, ...poolsB, ...poolsBase]) {
+    if (hasLiquidity(routePool)) {
+      routePoolMap.set(newRoutePoolKey(routePool.tokenA, routePool.tokenB, routePool.poolType), routePool);
+    }
+  }
+
+  // 1 hop
+  for (const baseToken of baseTokens) {
+    if (baseToken === tokenIn || baseToken === tokenOut) continue;
+
+    const poolA1 = routePoolMap.get(newRoutePoolKey(tokenIn, baseToken, 1));
+    const poolA2 = routePoolMap.get(newRoutePoolKey(tokenIn, baseToken, 2));
+
+    const poolB1 = routePoolMap.get(newRoutePoolKey(baseToken, tokenOut, 1));
+    const poolB2 = routePoolMap.get(newRoutePoolKey(baseToken, tokenOut, 2));
+
+    paths.push(
+      ...getPathsWith1Hop(
+        [poolA1, poolA2].filter(Boolean) as RoutePool[],
+        [poolB1, poolB2].filter(Boolean) as RoutePool[],
+        tokenIn,
+        baseToken
+      )
+    );
+  }
+
+  // 2 hops
+  for (const baseToken1 of baseTokens) {
+    if (baseToken1 === tokenIn || baseToken1 === tokenOut) continue;
+    const poolA1 = routePoolMap.get(newRoutePoolKey(tokenIn, baseToken1, 1));
+    const poolA2 = routePoolMap.get(newRoutePoolKey(tokenIn, baseToken1, 2));
+
+    for (const baseToken2 of baseTokens) {
+      if (baseToken2 === tokenIn || baseToken2 === tokenOut || baseToken2 === baseToken1) continue;
+
+      const poolBase1 = routePoolMap.get(newRoutePoolKey(baseToken1, baseToken2, 1));
+      const poolBase2 = routePoolMap.get(newRoutePoolKey(baseToken1, baseToken2, 2));
+
+      const poolB1 = routePoolMap.get(newRoutePoolKey(baseToken2, tokenOut, 1));
+      const poolB2 = routePoolMap.get(newRoutePoolKey(baseToken2, tokenOut, 2));
+
+      paths.push(
+        ...getPathsWith2Hops(
+          [poolA1, poolA2].filter(Boolean) as RoutePool[],
+          [poolBase1, poolBase2].filter(Boolean) as RoutePool[],
+          [poolB1, poolB2].filter(Boolean) as RoutePool[],
+          tokenIn,
+          baseToken1,
+          baseToken2
+        )
+      );
+    }
+  }
+
+  return paths;
+}
+
+const granularity = 10; // div to 10 parts
+
+async function splitAmount(amount: BigNumber, parts: number): Promise<BigNumber[][]> {
```

Comment:
Is it a good idea that we put all the logic from `router_utils.ts` in a single file so that we would not mix the logic together? It would also be easier to mention the source link in the file header.

Reply:
[Oral Discussion]